Provider crash due to conversion panic

Hi,

We're running into what seems to be a provider bug, but we're unsure as of yet.

Resources it seems to affect;

- betteruptime_monitor (primarily, the rest seems to crash due to the crash itself - see logs)
- betteruptime_google_monitoring_integration
- betteruptime_status_page_section

I have not been able to reproduce it outside of our production environment yet, but also cannot share those details. These kinds of errors can happen in case of a lack of memory, we have not validated this yet. It would be unusual, given we run on a 32GB RAM environment that is very lean. I'm aware this is limited information to do anything with, but wanted to report it regardless and keep you in the loop in case we find a root cause / fix. If this is something that someone else has seen / if this is a known issue - I'd eagerly would want to know.

```
2024-06-25 15:16:18.162 CEST
The plugin.(*GRPCProvider).ValidateResourceConfig request was cancelled.
2024-06-25 15:16:18.186 CEST
Error: Request cancelled


2024-06-25 15:16:18.327 CEST
The plugin encountered an error, and failed to respond to the
2024-06-25 15:16:18.327 CEST
plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain
2024-06-25 15:16:18.327 CEST
more details.
2024-06-25 15:16:18.328 CEST
2024-06-25 15:16:18.328 CEST
Error: Plugin did not respond


2024-06-25 15:16:18.384 CEST
panic: interface conversion: interface {} is nil, not map[string]interface {}
2024-06-25 15:16:18.384 CEST
2024-06-25 15:16:18.384 CEST
goroutine 92 [running]:
2024-06-25 15:16:18.384 CEST
github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider.loadRequestHeaders(0xc0000da680, 0xc0001cc650)
2024-06-25 15:16:18.384 CEST
	github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider/resource_monitor.go:574 +0x99f
2024-06-25 15:16:18.384 CEST
github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider.monitorCreate(0xda3c88, 0xc000756120, 0xc0000da680, 0xc08e00, 0xc000272bc0, 0xc00065c6d0, 0x834e2a, 0xc000600ba0)
2024-06-25 15:16:18.384 CEST
	github.com/BetterStackHQ/terraform-provider-better-uptime/internal/provider/resource_monitor.go:514 +0x117
2024-06-25 15:16:18.384 CEST
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0xc00043c9a0, 0xda3c18, 0xc0002735c0, 0xc0000da680, 0xc08e00, 0xc000272bc0, 0x0, 0x0, 0x0)
2024-06-25 15:16:18.384 CEST
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:330 +0x17f
2024-06-25 15:16:18.384 CEST
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00043c9a0, 0xda3c18, 0xc0002735c0, 0xc0008181c0, 0xc000600ba0, 0xc08e00, 0xc000272bc0, 0x0, 0x0, 0x0, ...)
2024-06-25 15:16:18.384 CEST
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/resource.go:456 +0x67b
2024-06-25 15:16:18.384 CEST
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc00000c0c0, 0xda3c18, 0xc0002735c0, 0xc0003dd8b0, 0xc0002735c0, 0xc70f00, 0xc0006b8300)
2024-06-25 15:16:18.384 CEST
	github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.0/helper/schema/grpc_provider.go:955 +0x8ef
2024-06-25 15:16:18.384 CEST
github.com/hashicorp/terraform-plugin-go/tfprotov5/server.(*server).ApplyResourceChange(0xc000600220, 0xda3cc0, 0xc0002735c0, 0xc000818070, 0xc000600220, 0xc0006b8390, 0xc00075eba0)
2024-06-25 15:16:18.384 CEST
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/server/server.go:332 +0xb5
2024-06-25 15:16:18.384 CEST
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0xc70f00, 0xc000600220, 0xda3cc0, 0xc0006b8390, 0xc00038a3c0, 0x0, 0xda3cc0, 0xc0006b8390, 0xc00016ce00, 0x6db)
2024-06-25 15:16:18.384 CEST
	github.com/hashicorp/terraform-plugin-go@v0.3.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
2024-06-25 15:16:18.384 CEST
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000f2540, 0xdaabb8, 0xc00038fb00, 0xc00020a000, 0xc0001227b0, 0x11cdc60, 0x0, 0x0, 0x0)
2024-06-25 15:16:18.384 CEST
	google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
2024-06-25 15:16:18.384 CEST
google.golang.org/grpc.(*Server).handleStream(0xc0000f2540, 0xdaabb8, 0xc00038fb00, 0xc00020a000, 0x0)
2024-06-25 15:16:18.384 CEST
	google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
2024-06-25 15:16:18.384 CEST
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000182410, 0xc0000f2540, 0xdaabb8, 0xc00038fb00, 0xc00020a000)
2024-06-25 15:16:18.384 CEST
	google.golang.org/grpc@v1.32.0/server.go:859 +0xab
2024-06-25 15:16:18.384 CEST
created by google.golang.org/grpc.(*Server).serveStreams.func1
2024-06-25 15:16:18.384 CEST
	google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd
2024-06-25 15:16:18.384 CEST
2024-06-25 15:16:18.384 CEST
Error: The terraform-provider-better-uptime_v0.11.1 plugin crashed!
2024-06-25 15:16:18.384 CEST
2024-06-25 15:16:18.384 CEST
This is always indicative of a bug within the plugin. It would be immensely
2024-06-25 15:16:18.384 CEST
helpful if you could report the crash with the plugin's maintainers so that it
2024-06-25 15:16:18.384 CEST
can be fixed. The output above should help diagnose the issue.
```

--- a/Makefile
+++ b/Makefile
@@ -8,7 +8,7 @@ GOLANGCI_LINT := golangci-lint run --disable-all \
 	-E staticcheck \
 	-E typecheck \
 	-E unused
-VERSION := 0.11.3
+VERSION := 0.11.7
 .PHONY: test build
 
 help:

--- a/examples/advanced/main.tf
+++ b/examples/advanced/main.tf
@@ -37,6 +37,12 @@ resource "betteruptime_monitor" "status" {
   url              = "https://example.com"
   monitor_type     = "status"
   monitor_group_id = betteruptime_monitor_group.this.id
+  request_headers = [
+    {
+      "name" : "X-For-Status-Page",
+      "value" : "https://${betteruptime_status_page.this.subdomain}.betteruptime.com"
+    },
+  ]
 }
 
 resource "betteruptime_monitor" "dns" {

--- a/examples/advanced/versions.tf
+++ b/examples/advanced/versions.tf
@@ -3,7 +3,7 @@ terraform {
   required_providers {
     betteruptime = {
       source  = "BetterStackHQ/better-uptime"
-      version = ">= 0.8.0"
+      version = ">= 0.11.6"
     }
   }
 }

--- a/internal/provider/resource_aws_cloudwatch_integration.go
+++ b/internal/provider/resource_aws_cloudwatch_integration.go
@@ -152,7 +152,9 @@ func awsCloudWatchIntegrationCreate(ctx context.Context, d *schema.ResourceData,
 	var in awsCloudWatchIntegration
 	for _, e := range awsCloudWatchIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -193,7 +195,9 @@ func awsCloudWatchIntegrationUpdate(ctx context.Context, d *schema.ResourceData,
 	for _, e := range awsCloudWatchIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_aws_cloudwatch_integration.go
+++ b/internal/provider/resource_aws_cloudwatch_integration.go
@@ -97,8 +97,9 @@ func newAwsCloudWatchIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/aws-cloudwatch-integrations/",
-		Schema:      awsCloudWatchIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/aws-cloudwatch-integrations/",
+		Schema:        awsCloudWatchIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_azure_integration.go
+++ b/internal/provider/resource_azure_integration.go
@@ -152,7 +152,9 @@ func azureIntegrationCreate(ctx context.Context, d *schema.ResourceData, meta in
 	var in azureIntegration
 	for _, e := range azureIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -193,7 +195,9 @@ func azureIntegrationUpdate(ctx context.Context, d *schema.ResourceData, meta in
 	for _, e := range azureIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_azure_integration.go
+++ b/internal/provider/resource_azure_integration.go
@@ -97,8 +97,9 @@ func newAzureIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/azure-integrations/",
-		Schema:      azureIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/azure-integrations/",
+		Schema:        azureIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_datadog_integration.go
+++ b/internal/provider/resource_datadog_integration.go
@@ -103,8 +103,9 @@ func newDatadogIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/datadog-integrations/",
-		Schema:      datadogIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/datadog-integrations/",
+		Schema:        datadogIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_datadog_integration.go
+++ b/internal/provider/resource_datadog_integration.go
@@ -160,7 +160,9 @@ func datadogIntegrationCreate(ctx context.Context, d *schema.ResourceData, meta 
 	var in datadogIntegration
 	for _, e := range datadogIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -201,7 +203,9 @@ func datadogIntegrationUpdate(ctx context.Context, d *schema.ResourceData, meta 
 	for _, e := range datadogIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_google_monitoring_integration.go
+++ b/internal/provider/resource_google_monitoring_integration.go
@@ -97,8 +97,9 @@ func newGoogleMonitoringIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/google-monitoring-integrations/",
-		Schema:      googleMonitoringIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/google-monitoring-integrations/",
+		Schema:        googleMonitoringIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_google_monitoring_integration.go
+++ b/internal/provider/resource_google_monitoring_integration.go
@@ -152,7 +152,9 @@ func googleMonitoringIntegrationCreate(ctx context.Context, d *schema.ResourceDa
 	var in googleMonitoringIntegration
 	for _, e := range googleMonitoringIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -193,7 +195,9 @@ func googleMonitoringIntegrationUpdate(ctx context.Context, d *schema.ResourceDa
 	for _, e := range googleMonitoringIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_grafana_integration.go
+++ b/internal/provider/resource_grafana_integration.go
@@ -97,8 +97,9 @@ func newGrafanaIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/grafana-integrations/",
-		Schema:      grafanaIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/grafana-integrations/",
+		Schema:        grafanaIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_grafana_integration.go
+++ b/internal/provider/resource_grafana_integration.go
@@ -152,7 +152,9 @@ func grafanaIntegrationCreate(ctx context.Context, d *schema.ResourceData, meta 
 	var in grafanaIntegration
 	for _, e := range grafanaIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -193,7 +195,9 @@ func grafanaIntegrationUpdate(ctx context.Context, d *schema.ResourceData, meta 
 	for _, e := range grafanaIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_metadata.go
+++ b/internal/provider/resource_metadata.go
@@ -116,7 +116,9 @@ func metadataCreate(ctx context.Context, d *schema.ResourceData, meta interface{
 	var in metadata
 	for _, e := range metadataRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -157,7 +159,9 @@ func metadataUpdate(ctx context.Context, d *schema.ResourceData, meta interface{
 	for _, e := range metadataRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_metadata.go
+++ b/internal/provider/resource_metadata.go
@@ -69,8 +69,9 @@ func newMetadataResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/metadata/",
-		Schema:      metadataSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/metadata/",
+		Schema:        metadataSchema,
 	}
 }
 

--- a/internal/provider/resource_monitor.go
+++ b/internal/provider/resource_monitor.go
@@ -397,8 +397,9 @@ func newMonitorResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/monitors/",
-		Schema:      monitorSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/monitors/",
+		Schema:        monitorSchema,
 	}
 }
 
@@ -564,6 +565,37 @@ func monitorUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}
 
 func monitorDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	return resourceDelete(ctx, meta, fmt.Sprintf("/api/v2/monitors/%s", url.PathEscape(d.Id())))
+}
+
+func validateRequestHeaders(ctx context.Context, diff *schema.ResourceDiff, v interface{}) error {
+	if headers, ok := diff.GetOk("request_headers"); ok {
+		for _, header := range headers.([]interface{}) {
+			headerMap := header.(map[string]interface{})
+			if err := validateRequestHeader(headerMap); err != nil {
+				return fmt.Errorf("Invalid request header %v: %v", headerMap, err)
+			}
+		}
+	}
+	return nil
+}
+
+func validateRequestHeader(header map[string]interface{}) error {
+	name, nameOk := header["name"].(string)
+	value, valueOk := header["value"].(string)
+
+	if !nameOk || name == "" {
+		return fmt.Errorf("must contain 'name' key with a non-empty string value")
+	}
+
+	if !valueOk || value == "" {
+		return fmt.Errorf("must contain 'value' key with a non-empty string value")
+	}
+
+	if len(header) != 2 {
+		return fmt.Errorf("must only contain 'name' and 'value' keys")
+	}
+
+	return nil
 }
 
 func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interface{}) {

--- a/internal/provider/resource_monitor.go
+++ b/internal/provider/resource_monitor.go
@@ -512,7 +512,9 @@ func monitorCreate(ctx context.Context, d *schema.ResourceData, meta interface{}
 	var in monitor
 	for _, e := range monitorRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -553,7 +555,9 @@ func monitorUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}
 	for _, e := range monitorRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}
@@ -580,6 +584,11 @@ func validateRequestHeaders(ctx context.Context, diff *schema.ResourceDiff, v in
 }
 
 func validateRequestHeader(header map[string]interface{}) error {
+	if len(header) == 0 {
+		// Headers with calculated fields that are not known at the time will be passed as empty maps, ignore them
+		return nil
+	}
+
 	name, nameOk := header["name"].(string)
 	value, valueOk := header["value"].(string)
 
@@ -598,7 +607,7 @@ func validateRequestHeader(header map[string]interface{}) error {
 	return nil
 }
 
-func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interface{}) {
+func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interface{}) error {
 	x := receiver
 	v := d.Get("request_headers")
 	var t []map[string]interface{}
@@ -613,6 +622,14 @@ func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interfac
 	for _, v := range old.([]interface{}) {
 		header := v.(map[string]interface{})
 		foundHeader := findRequestHeader(&t, &header)
+
+		// Validation at apply time, empty map is considered invalid (fields should be known at this point)
+		if len(header) == 0 {
+			return fmt.Errorf("Invalid request header %v: map cannot be empty", header)
+		}
+		if err := validateRequestHeader(header); err != nil {
+			return fmt.Errorf("Invalid request header %v: %v", header, err)
+		}
 
 		var hasId bool // Check if the found header already has an ID
 		if foundHeader == nil {
@@ -630,6 +647,8 @@ func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interfac
 	}
 
 	*x = &t
+
+	return nil
 }
 
 func findRequestHeader(headers *[]map[string]interface{}, header *map[string]interface{}) *map[string]interface{} {

--- a/internal/provider/resource_monitor.go
+++ b/internal/provider/resource_monitor.go
@@ -613,6 +613,22 @@ func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interfac
 	var t []map[string]interface{}
 	for _, v := range v.([]interface{}) {
 		header := v.(map[string]interface{})
+
+		// Validation at apply time, empty map is considered invalid (fields should be known at this point)
+		if len(header) == 0 {
+			return fmt.Errorf("Invalid request header %v: map cannot be empty", header)
+		}
+		// Headers can have ID at apply time, temporarily remove it before validation and reattach it afterwards
+		id, idPresent := header["id"]
+		delete(header, "id")
+		err := validateRequestHeader(header)
+		if idPresent {
+			header["id"] = id
+		}
+		if err != nil {
+			return fmt.Errorf("Invalid request header %v: %v", header, err)
+		}
+
 		newHeader := map[string]interface{}{"name": header["name"], "value": header["value"]}
 		t = append(t, newHeader)
 	}
@@ -622,14 +638,6 @@ func loadRequestHeaders(d *schema.ResourceData, receiver **[]map[string]interfac
 	for _, v := range old.([]interface{}) {
 		header := v.(map[string]interface{})
 		foundHeader := findRequestHeader(&t, &header)
-
-		// Validation at apply time, empty map is considered invalid (fields should be known at this point)
-		if len(header) == 0 {
-			return fmt.Errorf("Invalid request header %v: map cannot be empty", header)
-		}
-		if err := validateRequestHeader(header); err != nil {
-			return fmt.Errorf("Invalid request header %v: %v", header, err)
-		}
 
 		var hasId bool // Check if the found header already has an ID
 		if foundHeader == nil {

--- a/internal/provider/resource_monitor_test.go
+++ b/internal/provider/resource_monitor_test.go
@@ -6,6 +6,7 @@ import (
 	"io"
 	"net/http"
 	"net/http/httptest"
+	"regexp"
 	"strconv"
 	"sync/atomic"
 	"testing"
@@ -193,7 +194,7 @@ func TestResourceMonitorWithHeaders(t *testing.T) {
 					resource.TestCheckResourceAttr("betteruptime_monitor.this", "request_headers.#", "2"),
 				),
 			},
-			// Step 2 - remove the first header.
+			// Step 3 - remove the first header.
 			{
 				Config: fmt.Sprintf(`
 				provider "betteruptime" {
@@ -217,6 +218,90 @@ func TestResourceMonitorWithHeaders(t *testing.T) {
 					resource.TestCheckResourceAttr("betteruptime_monitor.this", "request_headers.0.value", "test-2"),
 					resource.TestCheckNoResourceAttr("betteruptime_monitor.this", "request_headers.1.name"),
 				),
+			},
+			// Step 4 - invalid header with empty name.
+			{
+				Config: fmt.Sprintf(`
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "%s"
+					monitor_type = "%s"
+					request_headers = [
+						{
+							name  = ""
+							value = "test"
+						}
+					]
+				}
+				`, url, monitorType),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile(`Invalid request header map\[name: value:test\]: must contain 'name' key with a non-empty string value`),
+			},
+			// Step 5 - invalid header with empty value.
+			{
+				Config: fmt.Sprintf(`
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "%s"
+					monitor_type = "%s"
+					request_headers = [
+						{
+							name  = "X-TEST"
+							value = ""
+						}
+					]
+				}
+				`, url, monitorType),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile(`Invalid request header map\[name:X-TEST value:\]: must contain 'value' key with a non-empty string value`),
+			},
+			// Step 6 - invalid header with extra keys.
+			{
+				Config: fmt.Sprintf(`
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "%s"
+					monitor_type = "%s"
+					request_headers = [
+						{
+							name  = "X-TEST"
+							value = "test"
+							extra = "invalid"
+						}
+					]
+				}
+				`, url, monitorType),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile(`Invalid request header map\[extra:invalid name:X-TEST value:test\]: must only contain 'name' and 'value' keys`),
+			},
+			// Step 7 - invalid header with incorrect format.
+			{
+				Config: fmt.Sprintf(`
+				provider "betteruptime" {
+					api_token = "foo"
+				}
+
+				resource "betteruptime_monitor" "this" {
+					url          = "%s"
+					monitor_type = "%s"
+					request_headers = [
+						{
+							"X-TEST" = "test"
+						}
+					]
+				}
+				`, url, monitorType),
+				PlanOnly:    true,
+				ExpectError: regexp.MustCompile(`Invalid request header map\[X-TEST:test\]: must contain 'name' key with a non-empty string value`),
 			},
 		},
 	})

--- a/internal/provider/resource_new_relic_integration.go
+++ b/internal/provider/resource_new_relic_integration.go
@@ -103,8 +103,9 @@ func newNewRelicIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/new-relic-integrations/",
-		Schema:      newRelicIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/new-relic-integrations/",
+		Schema:        newRelicIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_new_relic_integration.go
+++ b/internal/provider/resource_new_relic_integration.go
@@ -160,7 +160,9 @@ func newRelicIntegrationCreate(ctx context.Context, d *schema.ResourceData, meta
 	var in newRelicIntegration
 	for _, e := range newRelicIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -201,7 +203,9 @@ func newRelicIntegrationUpdate(ctx context.Context, d *schema.ResourceData, meta
 	for _, e := range newRelicIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_pagerduty_integration.go
+++ b/internal/provider/resource_pagerduty_integration.go
@@ -48,8 +48,9 @@ func newPagerdutyIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/pagerduty-integrations/",
-		Schema:      pagerdutyIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/pagerduty-integrations/",
+		Schema:        pagerdutyIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_pagerduty_integration.go
+++ b/internal/provider/resource_pagerduty_integration.go
@@ -87,7 +87,9 @@ func pagerdutyIntegrationCreate(ctx context.Context, d *schema.ResourceData, met
 	var in pagerdutyIntegration
 	for _, e := range pagerdutyIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -128,7 +130,9 @@ func pagerdutyIntegrationUpdate(ctx context.Context, d *schema.ResourceData, met
 	for _, e := range pagerdutyIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_prometheus_integration.go
+++ b/internal/provider/resource_prometheus_integration.go
@@ -97,8 +97,9 @@ func newPrometheusIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/prometheus-integrations/",
-		Schema:      prometheusIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/prometheus-integrations/",
+		Schema:        prometheusIntegrationSchema,
 	}
 }
 

--- a/internal/provider/resource_prometheus_integration.go
+++ b/internal/provider/resource_prometheus_integration.go
@@ -152,7 +152,9 @@ func prometheusIntegrationCreate(ctx context.Context, d *schema.ResourceData, me
 	var in prometheusIntegration
 	for _, e := range prometheusIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -193,7 +195,9 @@ func prometheusIntegrationUpdate(ctx context.Context, d *schema.ResourceData, me
 	for _, e := range prometheusIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_splunk_on_call_integration.go
+++ b/internal/provider/resource_splunk_on_call_integration.go
@@ -87,7 +87,9 @@ func splunkOnCallIntegrationCreate(ctx context.Context, d *schema.ResourceData, 
 	var in splunkOnCallIntegration
 	for _, e := range splunkOnCallIntegrationRef(&in) {
 		if e.k == "request_headers" {
-			loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+			if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+				return diag.FromErr(err)
+			}
 		} else {
 			load(d, e.k, e.v)
 		}
@@ -128,7 +130,9 @@ func splunkOnCallIntegrationUpdate(ctx context.Context, d *schema.ResourceData, 
 	for _, e := range splunkOnCallIntegrationRef(&in) {
 		if d.HasChange(e.k) {
 			if e.k == "request_headers" {
-				loadRequestHeaders(d, e.v.(**[]map[string]interface{}))
+				if err := loadRequestHeaders(d, e.v.(**[]map[string]interface{})); err != nil {
+					return diag.FromErr(err)
+				}
 			} else {
 				load(d, e.k, e.v)
 			}

--- a/internal/provider/resource_splunk_on_call_integration.go
+++ b/internal/provider/resource_splunk_on_call_integration.go
@@ -48,8 +48,9 @@ func newSplunkOnCallIntegrationResource() *schema.Resource {
 		Importer: &schema.ResourceImporter{
 			StateContext: schema.ImportStatePassthroughContext,
 		},
-		Description: "https://betterstack.com/docs/uptime/api/splunk-on-call-integrations/",
-		Schema:      splunkOnCallIntegrationSchema,
+		CustomizeDiff: validateRequestHeaders,
+		Description:   "https://betterstack.com/docs/uptime/api/splunk-on-call-integrations/",
+		Schema:        splunkOnCallIntegrationSchema,
 	}
 }
 

